substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java

```diff
@@ -186,31 +265,66 @@ private static boolean wasAlreadyInCanonicalForm(String resourceName, String can
         return resourceName.equals(canonicalResourceName) || removeTrailingSlash(resourceName).equals(canonicalResourceName);
     }
 
-    public static ResourceStorageEntry get(String name) {
-        return get(null, name);
+    public static Object get(String name, boolean throwOnMissing) {
+        return get(null, name, throwOnMissing);
     }
 
-    public static ResourceStorageEntry get(String moduleName, String resourceName) {
+    /**
+     * If {@code throwOnMissing} is false, we have to distinguish an entry that was in the metadata
+     * from one that was not, so the caller can correctly throw the
+     * {@link MissingResourceMetadataException}. This is needed because different modules can be
+     * tried on the same resource name, causing an unexpected exception if we throw directly.
+     */
+    public static Object get(String moduleName, String resourceName, boolean throwOnMissing) {
         String canonicalResourceName = toCanonicalForm(resourceName);
-        ResourceStorageEntry entry = singleton().resources.get(Pair.create(moduleName, canonicalResourceName));
+        Object entry = singleton().resources.get(Pair.create(moduleName, canonicalResourceName));
         if (entry == null) {
+            if (MissingResourceMetadataException.Options.ThrowMissingMetadataExceptions.getValue()) {
+                for (Pair<String, Pattern> pattern : singleton().excludePatterns) {
+                    if (Objects.equals(moduleName, pattern.getLeft()) && (pattern.getRight().matcher(resourceName).matches() || pattern.getRight().matcher(canonicalResourceName).matches())) {
+                        return missingMetaData(resourceName, throwOnMissing);
+                    }
+                }
+                for (Pair<String, Pattern> pattern : singleton().includePatterns) {
+                    if (Objects.equals(moduleName, pattern.getLeft()) && (pattern.getRight().matcher(resourceName).matches() || pattern.getRight().matcher(canonicalResourceName).matches())) {
+                        return null;
+                    }
+                }
+                return missingMetaData(resourceName, throwOnMissing);
+            } else {
+                return null;
+            }
+        }
+        if (entry instanceof IOException) {
+            throw new RuntimeException((IOException) entry);
```

Comment:
Shouldn't the `IOException` be thrown directly?

Reply:
Do you mean at build time? I don't think we can simply throw the `IOException` itself, as we need to either handle it, or add it in the method signature.